packages/tldraw/src/lib/shapes/shared/useAsset.ts

```diff
@@ -0,0 +1,25 @@
+import { TLAssetId, debounce, useEditor, useValue } from '@tldraw/editor'
+import { useEffect, useRef, useState } from 'react'
+
+/** @internal */
+export function useAsset(assetId: TLAssetId | null) {
+	const editor = useEditor()
+	const [url, setUrl] = useState('')
+	const asset = assetId ? editor.getAsset(assetId) : null
+
+	// We debounce the zoom level to reduce the number of times we fetch a new image and,
+	// more importantly, to not cause zooming in and out to feel janky.
+	const [debouncedZoom, setDebouncedZoom] = useState(editor.getZoomLevel())
+	const zoomUpdater = useRef(debounce((zoom: number) => setDebouncedZoom(zoom), 500))
+	useValue('zoom level', () => zoomUpdater.current(editor.getZoomLevel()), [editor])
```

Comment:
The way this is structured is a bit of a react anti-pattern:
- we create a new debounce function every render
- we don't cancel debounced calls after the function unmounts, resulting in set-state calls on unmounted components

im vaguely surprised we don't have a `useDebouncedValue` type utility hook already, but if you wanted to add one you could do something like:
```tsx
function useDebouncedValue<T>(value: T, ms: number): T {
    const [debouncedValue, setDebouncedValue] = useState(value)

    useEffect(() => {
        const timer = setTimeout(() => {
            setDebouncedValue(value)
        }, ms)
        return () => clearTimeout(timer)
    }, [value, ms])

    return value
}

// usage:
const rawZoomLevel = useValue(...)
const debouncedZoomLevel = useDebouncedValue(rawZoomLevel, 500)
```

Reply:
> we create a new debounce function every render

correct me if i'm wrong here but the signature of useRef is: `useRef(initialValue)` - the argument is ignored after initial render... 🤔

Reply:
yup! it doesn't get used, but we do do the work of creating a new debounced function. that first bullet isn't at all a big deal other than creating garbage. i think i saw it has a bit of a smell that we were doing side-effect-y things in the render phase though (which is why we don't have cancelling rn)

Reply:
i see, i see, gotcha